src/components/pokemon/shared/evolution-tree.ts

```diff
@@ -11,10 +11,16 @@ class EvolutionTree implements Identifyable {
      */
     public name: string;
     public id: string;
+    public candyCost: number;
+    public evolutionItem: Identifyable;
+    public kmBuddyDistanceRequirement: number;
     public futureEvolutions: EvolutionTree[] = [];
-    public constructor(name: string, id: string, futureEvolutions: EvolutionTree[] = []) {
+    public constructor(name: string, id: string, candyCost: number = undefined, evolutionItem: Identifyable = undefined, kmBuddyDistanceRequirement: number = undefined, futureEvolutions: EvolutionTree[] = []) {
```

Comment:
We should refactor this. If I can't come up with an easy solution we'll definitely have to create an issue for that

Reply:
Each element in the tree should have all the required information, right? Therefore, those need to be passed in the constructor (unless you want to have a dictionary somewhere and try to obtain that information from previous evolution's `nextEvolutionBranches`).

Again, it was convenient to just pass it directly.

Reply:
However, the whole design of `futureEvolutions` creates a lot of data duplication. Maybe instead of storing the information here, there could be a helper function to just obtain it based on the `nextEvolutionBranches` data?

Reply:
```
    "futureEvolutions": {
        "futureEvolutions": [],
        "name": "Kingdra",
        "id": "KINGDRA",
        "candyCost": 100,
        "evolutionItem": {
            "id": "ITEM_DRAGON_SCALE",
            "name": "Dragon Scale"
         }
    }
```

could become:
```
    "futureEvolutions": [],
    "candyCost": 100,
    "evolutionItem": {
        "id": "ITEM_DRAGON_SCALE",
        "name": "Dragon Scale"
    }
```

So that `futureEvolutions` follows `pastEvolutions` format and becomes `nextEvolutionBranches`-like with full evo chain, while mon's evo requirements become their direct "attributes"

Reply:
Future Evolutions and Past Evolutions, at present, make specific information more accessible (ie, which pokemon are in the same evolution family) by duplicating it. This saves on code duplication and processing overhead elsewhere. We can't use *just* a simple array for future evolutions because there needs to be a single data structure that represents both Bulbasaur -> Venosaur and the Eevee family. There's no way to represent the full complexity that pokemon evolutions could have with a 1-dimensional collection.

That's why there's some awkwardness with the `futureEvolutions` property; arrays of recursive objects are the only way I know of to make it clear how the tree branches when. In past evolutions, we can (for now) get away with a simple line, because we don't care about branches not taken. In the future, if there are evolutions that merge, the structure of past evolutions may need to be improved.

I'm not against adding evolution costs to the `futureEvolutions` tree directly. I'm not sure whether it is clear that you're talking about incremental cost or total cost, but that's a nuance that can be argued about. Once you add that data, though, you need to recognize the graph-like data structure that you are serializing into a hierarchical data format. Nothing is wrong with how the format works in this scenario, as long as everyone involved recognizes the complexity.

Reply:
But it does contain the evolution costs inside the doubled `futureEvolutions`. It's just that the "main" evolution cost could be moved outside and then there won't be need to involve the weirdness. See my comment on https://github.com/BrunnerLivio/pokemongo-json-pokedex/issues/47